specification/dartLangSpec.tex

```diff
@@ -21253,39 +21261,2076 @@ \subsubsection{Informal Subtype Rule Descriptions}
+
+  \noindent
+  \code{$T_2$\,\FUNCTION<$X_1$\,\EXTENDS\,$B_{21}$,\,\ldots,\,$X_m$\,%
+    \EXTENDS\,$B_{2m}$>($P_{21}$,\,\ldots,\,$P_{2l}$)}
+
+  \noindent
+  such that each $B_{1i}$ and $B_{2i}$ are types with the same canonical syntax,
+  and both have the same number of required positional parameters.
+  Let $q$ be $\metavar{min}(k, l)$,
+  let $T_3$ be \UpperBoundType{$T_1$}{$T_2$},
+  let $B_{3i}$ be $B_{1i}$, and
+  let $P_{3i}$ be \LowerBoundType{$P_{1i}$}{$P_{2i}$}.
+  Then \DefEquals{\UpperBoundType{$U_1$}{$U_2$}}{%
+    \code{$T_3$\,\FUNCTION<$X_1$\,\EXTENDS\,$B_{31}$,\,\ldots,\,$X_m$\,%
+      \EXTENDS\,$B_{3m}$>($P_{31}$,\,\ldots,\,$P_{3q}$)}}.
+
+  \commentary{%
+    This case includes non-generic function types by allowing $m$ to be zero.%
+  }
+\item
+  Let $U_1$ and $U_2$ be, respectively,
+
+  \noindent
+  \code{$T_1$\,\FUNCTION<$X_1$\,\EXTENDS\,$B_{11}$,\,\ldots,\,$X_m$\,%
+    \EXTENDS\,$B_{1m}$>($P_{11}$,\,\ldots,\,$P_{1k}$,\,$\metavar{Named}_1$)}
+
+  \noindent
+  \code{$T_2$\,\FUNCTION<$X_1$\,\EXTENDS\,$B_{21}$,\,\ldots,\,$X_m$\,%
+    \EXTENDS\,$B_{2m}$>($P_{21}$,\,\ldots,\,$P_{2k}$,\,$\metavar{Named}_2$)}
+
+  \noindent
+  where $\metavar{Named}_j$ declares a non-empty set of named parameters
+  with names $\metavar{NamesOfNamed}_j$, $j \in 1 .. 2$,
+  and consider the case where the following is satisfied:
+
+  \begin{itemize}
+  \item Each $B_{1i}$ and $B_{2i}$ are types with the same canonical syntax.
+  \item For each required entry named $n$ in $\metavar{Named}_1$,
+    $\metavar{Named}_2$ contains an entry named $n$
+    (\commentary{which may or may not be required}).
+  \item For each required entry named $n$ in $\metavar{Named}_2$,
+    $\metavar{Named}_1$ contains an entry named $n$
+    (\commentary{which may or may not be required}).
+  \end{itemize}
+
+  Then \DefEqualsNewline{\UpperBoundType{$U_1$}{$U_2$}}{%
+    \code{$T_3$\,\FUNCTION<$X_1$\,\EXTENDS\,$B_{31}$,\,\ldots,\,$X_m$\,%
+      \EXTENDS\,$B_{3m}$>($P_{31}$,\,\ldots,\,$P_{3k}$,\,$\metavar{Named}_3$)}},
+
+  \noindent
+  where:
+
+  \begin{itemize}
+  \item $T_3$ is \UpperBoundType{$T_1$}{$T_2$}.
+  \item $B_{3i}$ is $B_{1i}$.
+  \item $P_{3i}$ is \LowerBoundType{$P_{1i}$}{$P_{2i}$}.
+  \item $\metavar{Named}_3$ declares the set
+    of named parameter types with the names
+    $\metavar{NamesOfNamed}_1\cap\metavar{NamesOfNamed}_2$,
+    such that for each $P'$ in $\metavar{Named}_3$ with name $n$,
+    where $P'_1 \in \metavar{Named}_1$
+    and $P'_2 \in \metavar{Named}_2$
+    also have the name $n$,
+    $P'$ is \LowerBoundType{$P'_1$}{$P'_2$}.
+  \end{itemize}
+
+  \commentary{%
+    This case includes non-generic function types by allowing $m$ to be zero.%
+  }
+%%
+%% TODO(eernst), for review: Why do we not have a rule for
+%% \UpperBoundType{T1 Function(P1..Pm, [...])}{T2 Function(P1..Pk, {...}}}
+%% = T3 Function(R1..Rk), where the left operand has at least k parameters,
+%% plus the converse?
```

Comment:
It seems that this case is covered by treating the required and optional positional parameters in the positional-parameter case above uniformly. As I mentioned, the implementations seem to diverge from that. They expect the number of required positional parameters to be the same, and then just cut out the optional parameters in the upper bound.